rclcpp/include/rclcpp/qos.hpp

```diff
@@ -140,6 +148,31 @@ class RCLCPP_PUBLIC QoS
   const rmw_qos_profile_t &
   get_rmw_qos_profile() const;
 
+  /// Check if this QoS profile is compatible with another.
+  /**
+   * Two QoS profiles are compatible if a publisher and subcription
+   * using the QoS policies can communicate with each other.
+   *
+   * This QoS profile is assumed to be for a subscription.
+   *
+   * If there is an unknown policy value in either profile, then an exception is thrown.
+   *
+   * \param[in] publisher_qos: The QoS profile for a publisher.
```

Comment:
Why is this "publish" qos? Does the direction matter? I guess it might.

In which case maybe there should just be a free function like `check_qos_for_compatability(sub_qos, pub_qos, ...)`. I dunno.

It might be that int he future we need a separate QoS for publishers and for subscriptions, but for now it's odd to have this as a method, I think... I dunno, not a blocking thing, it just seems weird.

Reply:
The direction does matter. E.g.

- best effort publisher is **not** compatible with a reliable subscription
- reliable publisher **is** compatible with a best effort subscription

I arbitrarily presumed the caller would want to check a subscribers QoS against a fixed publisher QoS. But I like the idea of making this a free function.

Reply:
Refactored to be a free function: 92a1cef